server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
 
     // Tracks the latest known repository generation in a best-effort way to detect inconsistent listing of root level index-N blobs
     // and concurrent modifications.
-    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.EMPTY_REPO_GEN);
+    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.UNKNOWN_REPO_GEN);
 
     @Override
     public void getRepositoryData(ActionListener<RepositoryData> listener) {
-        ActionListener.completeWith(listener, () -> {
+        if (latestKnownRepoGen.get() == RepositoryData.CORRUPTED_REPO_GEN) {
+            listener.onFailure(corruptedStateException(null));
+            return;
+        }
+        if (bestEffortConsistency == false && latestKnownRepoGen.get() == RepositoryData.UNKNOWN_REPO_GEN) {
+            if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                // Non-master node does not load RepositoryData before master has at least loaded it once and thus initialized in the
+                // cluster state
+                listener.onFailure(new RepositoryException(metadata.name(), "Repository not initialized yet"));
```

Comment:
Non-master nodes only ever load `RepositoryData`  in `org.elasticsearch.index.shard.StoreRecovery#restore` (in which case master will have loaded `RepositoryData` before to verify that the snapshot to be restored actually exists in the repo) as far as I can see so in the real world we should never get here (except for some freak concurrency cases where master fails over concurrently).

Reply:
This actually becomes a non-issue thanks to your suggestion in https://github.com/elastic/elasticsearch/pull/49060#discussion_r357232276 since we can simply fall back to listing on the non-master nodes so long as the master hasn't initialized the latest generation in the CS